### ISSUE TITLE
Bug: [@typescript-eslint/no-unused-expressions] Error when using logical AND (&&) for null check

### ISSUE THREAD

**Issue by florianmutter** [typescript-eslint/typescript-eslint#9748]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.4.2&fileType=.ts&code=JYOwLgpgTgZghgYwgAgGIHt3IN4ChnIBGcUA-AFzIAUAlMgLwB8yAbusACa4C%2BuuC6EAGcwyGJkoYs9HL1zj0AOmJRkAMjVjMykrT4KdZRbSA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQHYHsBaWXWZRAE0MQA8AHaFZAS311Q0nwDNvIAacNkgABAC4BPOigDG0ZnTHVk8ZrjEB6AsVLkqtBk1bt0URNGj5okIQF8QtoA&tsconfig=&tokens=false

### Repro Code

```TypeScript
interface Foo {
  bar?: () => void
}

const foo: Foo = {}

foo.bar && foo.bar()

foo.bar?.()
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "no-unused-expressions": "off",
    "@typescript-eslint/no-unused-expressions": "error"
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
  }
}
```

### Expected Result

No error

### Actual Result

Line `foo.bar && foo.bar()` does not pass check

### Additional Info

This seems to have changed in version 8 but I could not find any reason why.

**Comment by bradzacher** [comment]
https://eslint.org/docs/latest/rules/no-unused-expressions#allowshortcircuit

Please checkout the rule docs

**Comment by florianmutter** [comment]
Thank you very much for the hint. I only checked https://typescript-eslint.io/rules/no-unused-expressions/ and did not check the eslint documentation
